tor/add_onion_test.go

```diff
@@ -2,50 +2,105 @@ package tor
 
 import (
 	"bytes"
+	"fmt"
 	"io/ioutil"
 	"path/filepath"
 	"testing"
+
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/lightningnetwork/lnd/keychain"
+)
+
+var (
+	testWalletPrivKey = []byte{
+		0x2b, 0xd8, 0x06, 0xc9, 0x7f, 0x0e, 0x00, 0xaf,
+		0x1a, 0x1f, 0xc3, 0x32, 0x8f, 0xa7, 0x63, 0xa9,
+		0x26, 0x97, 0x23, 0xc8, 0xdb, 0x8f, 0xac, 0x4f,
+		0x93, 0xaf, 0x71, 0xdb, 0x18, 0x6d, 0x6e, 0x90,
+	}
 )
 
-// TestOnionFile tests that the OnionFile implementation of the OnionStore
+type mockKeyRing struct {
```

Comment:
Should be in `lntest/mock`

Reply:
@Crypt-iQ, Thanks for the review. I was working on updating these mock keyrings to import the lntest/mock ones, but when I use them in the tor package it causes an import cycle. How would you recommend I resolve that?

Reply:
Could you post the import cycle message that you see or maybe a link to the branch you're using locally?

Reply:
```# github.com/lightningnetwork/lnd/tor
import cycle not allowed in test
package github.com/lightningnetwork/lnd/tor (test)
	imports github.com/lightningnetwork/lnd/lntest/mock
	imports github.com/lightningnetwork/lnd/chainntnfs
	imports github.com/lightningnetwork/lnd/channeldb
	imports github.com/lightningnetwork/lnd/channeldb/migration12
	imports github.com/lightningnetwork/lnd/lnwire
	imports github.com/lightningnetwork/lnd/tor

Reply:
I'll check it out - it seems like this might block this PR.

Reply:
@Crypt-iQ Sounds good. Let me know how you'd like me to proceed.

Reply:
I see the problem. `tor` is used as a library and imports `lntest/mock`.

Unfortunately `lntest/mock` imports a bunch of other unnecessary stuff.  We would need to move every file in `lntest/mock` into its own subpackage like `lntest/mock/keyring`. Then I think it should be possible for `tor` to import the mock key ring.

If we think about it from a `tor` library perspective where other projects can import it and use the functionality, then it can't have internal test dependencies on lnd. However, if we want to properly mock it, have it testable, and fuzzable within lnd, we need these mock dependencies. So I think the two approaches that wilmer and I have are at odds with each other. 

Not sure what to do about it ?
@wpaulino

Reply:
> If we think about it from a tor library perspective where other projects can import it and use the functionality, then it can't have internal test dependencies on lnd.

:+1:

It looks like we just need to move the `OnionFile` implementation and its accompanying tests into the main `lnd` package.

Reply:
@wpaulino Can you please elaborate a little more. I can try to separate the necessary pieces. Are you suggesting we ditch `lnd/tor` as a standalone package and move it into `lnd`? or rather, leave `lnd/tor` alone as much as possible and only move over the affected pieces?

I started looking into moving over the `OnionFile` pieces. The only affected function call is `NewOnionFile`. I could move just this function into `lnd`, but then we have a problem importing it into he `watchtower` package. We can't then import `NewOnionFile` from `lnd` into `watchtower` because that's a new circular import problem. Not to mention, I think the `NewOnionFile` function makes sense to stay in the `lnd/tor` package.

I'm not sure the best way to handle this. We could use the `lntest/mock` keyring everywhere except for the `lnd/tor` package and leave it as is? That way it's still a standalone package with standalone tests...

Reply:
> Can you please elaborate a little more. I can try to separate the necessary pieces. Are you suggesting we ditch lnd/tor as a standalone package and move it into lnd? or rather, leave lnd/tor alone as much as possible and only move over the affected pieces?

The idea is for the package to not include any `lnd` dependencies so that it can be used in other projects, but it also makes things "cleaner" from a code health perspective. This PR introduces some dependencies with the onion file encryption feature.

Do we run into any circular dependencies if we were to move the `OnionFile` definitions to the `lnencrypt` package? Or maybe it's best to create a new sub-package under `lnd/tor`?

Reply:
I see now that `tor` has no `lnd` dependencies. So then `lntest/mock` should not be used and it should have its own separate mock keyring. However, if `tor` is to be its own library, it should probably have its own repo or something because if this pattern replicates across the codebase I could see future headaches.

Reply:
I concur @Crypt-iQ. I think this might get a little easier too when watchtower is moved into it's own repo too. Because there's duplication there as well.

So how should I proceed here? In my opinion creating a mock keyring in the tor package and changing everything else to use `lntest/mock` would satisfy this..

Reply:
So the `tor` package would need its own mock key ring which can live in `tor`. Problem is that pesky `lnd` import for encrypting the private key.

Reply:
Ah drat. You're right. Ok so I have a proposal:

Add a new parameter to the `AddOnion` function of `returnKey`. If the `returnKey == true` then we don't write the key to disk in `AddOnion` and instead we return it as part of the `*OnionAddr` response. Then in the `server.go` we set `returnKey = tor.encryptkey`when calling `AddOnion`. If it returns the key then we encrypt and write to disk in `server.go`. If `tor.encryptkey == false` then everything works the same way it does today and the key in `*OnionAddr` is empty. This would remove any `lnd` dependencies in the `tor` package.

Reply:
@gkrizek your proposal sounds good to me if you want to go ahead and give it a try.